# Network generation, paths, connectivity
Sebastian Ruf 
with Emma K. Towlson, Louis Shekhtman, Michael Danziger and A.-Laszlo Barabasi

* We're going to write our own Erdos-Renyi generation algorithm and use NetworkX to demonstrate some of the concepts and algorithms in the textbook

* Open your IPython notebook and load `networkx`, `numpy`, and `matplotlib` as in the last class

In [1]:
import numpy as np
import networkx as nx
import matplotlib.pyplot as plt
import matplotlib as mpl
%matplotlib inline

* Also change the default plot settings so they aren't so heinous

In [2]:
# change defaults to be less ugly
mpl.rc('xtick', labelsize=14, color="#222222") 
mpl.rc('ytick', labelsize=14, color="#222222") 
mpl.rc('font', **{'family':'sans-serif','sans-serif':['Arial']})
mpl.rc('font', size=16)
mpl.rc('xtick.major', size=6, width=1)
mpl.rc('xtick.minor', size=3, width=1)
mpl.rc('ytick.major', size=6, width=1)
mpl.rc('ytick.minor', size=3, width=1)
mpl.rc('axes', linewidth=1, edgecolor="#222222", labelcolor="#222222")
mpl.rc('text', usetex=False, color="#222222")

# Basic data types in Python

## Primitives
Basic math operations are +, -, *, /. For exponentiation, use ** and for integer division use //

In [3]:
# integer
a = 4

# floating point
b = 100.1

# complex
c = 1 + 2j

# string
s = "Hello, world!"

# boolean
t = True
f = False

print(a**2)
print(b/a)
print(b//a)

16
25.025
25.0


## Lists
Basic ordered container. Supports O(1) indexing, adding to/removing from right side

In [4]:
x = [1, c, "c", 3.14]
print(x)

[1, (1+2j), 'c', 3.14]


In [5]:
# adding 
x.append("hello")
print("After adding 'hello' the list is")
print(x)

After adding 'hello' the list is
[1, (1+2j), 'c', 3.14, 'hello']


In [6]:
# removing
print("After removing 1 the list is")
x.remove(1)
print(x)

After removing 1 the list is
[(1+2j), 'c', 3.14, 'hello']


In [7]:
# testing membership
print(3.14 in x)

# testing index
print(x.index('c'))

True
1


In [8]:
# indexing and slicing. First is 0, last is -1
print("The third element is", x[2])
print("The last element is", x[-1])
print("The first two elements are", x[0:2])

The third element is 3.14
The last element is hello
The first two elements are [(1+2j), 'c']


In [9]:
# iterating
print("All the elements of x")
for elem in x:
    print(elem)

All the elements of x
(1+2j)
c
3.14
hello


In [10]:
# the built-in "enumerate is also handy sometimes
for index, elem in enumerate(x):
    print("Element number", index, "is", elem)

Element number 0 is (1+2j)
Element number 1 is c
Element number 2 is 3.14
Element number 3 is hello


## Tuples
Tuples are immutable lists, more or less

In [11]:
y = (1, 2, "a")
# that's it, no appending or removing allowed

In [12]:
# size
print(len(y))

3


In [13]:
# indexing, iterating and enumerating
print("The first value in y is", y[0])
print("The last two elements of y are", y[-2:])

print()
print("All the values of y")
for elem in y:
    print(elem)
    
print()
print("With indices")
for index, elem in enumerate(y):
    print("Item", index, "is", elem)

The first value in y is 1
The last two elements of y are (2, 'a')

All the values of y
1
2
a

With indices
Item 0 is 1
Item 1 is 2
Item 2 is a


## Sets
Unordered container that only allows unique elements

In [14]:
# creation
s = {1, 2, 3, 4}

# equivalently, s = set([1,2,3,4])

In [15]:
# testing membership
print(4 not in s)
print(1 in s)

False
True


In [16]:
# adding and deleting
s.add("Laszlo")
if 3 in s:
    s.remove(3)
print(s)

# adding a duplicate has no effect
s.add(1)
print(s)

{1, 2, 4, 'Laszlo'}
{1, 2, 4, 'Laszlo'}


In [17]:
# logical operations
s2 = set((3, "Laszlo", "Sebastian", 10, 3.14))

# union
print("Union of s and s2 is:")
print(s | s2)

print("Intersection of s and s2 is:")
print(s & s2)

print("Difference of s and s2 is:")
print(s2 - s)

Union of s and s2 is:
{1, 2, 3.14, 4, 3, 'Laszlo', 10, 'Sebastian'}
Intersection of s and s2 is:
{'Laszlo'}
Difference of s and s2 is:
{3, 10, 3.14, 'Sebastian'}


## Dictionaries
Dictionaries map a key to an associated value, much like a dictionary maps a word to a definition.
Values can be anything you want, keys can be almost anything (notable exception: lists)

In [40]:
# map name to zodiac sign
sign = {'Sebastian': 'Virgo', 'Nancy': 'Taurus'}
# equivalent: dict(Sebastian='Virgo', Nancy='Taurus')
# also equivalent: dict([('Sebastian', 'Virgo'), ('Nancy', 'Taurus')])

sign['Sebastian']

'Virgo'

In [41]:
# adding
sign['Alice'] = 'Libra'
print(sign)

{'Sebastian': 'Virgo', 'Nancy': 'Taurus', 'Alice': 'Libra'}


In [20]:
# number of keys
print("dictionary has", len(sign), "keys")

dictionary has 3 keys


In [42]:
# testing membership
print('Sebastian' in sign)
print('Taurus' in sign)

True
False


In [43]:
# removing
del sign['Sebastian']
print(sign)

{'Nancy': 'Taurus', 'Alice': 'Libra'}


In [44]:
# iterating over: keys, values, both (called "items")
print("Keys:")
for k in sign:
    print(k)
print()
print("Values:")
for v in sign.values():
    print(v)
print()
print("Both:")
for k, v in sign.items():
    print(k, "is a", v)

Keys:
Nancy
Alice

Values:
Taurus
Libra

Both:
Nancy is a Taurus
Alice is a Libra


## numpy arrays
* Another type you will use a lot is numpy's `ndarray`, which can represent vectors, matrices, etc.\
* A 1D ndarray (a vector) is like a list, but supports vectorized operations
* This is python's equivalent of the basic data type in MATLAB

In [24]:
# vectors and vector operations
# all are element-wise
a = np.array([1, 2, 3])
b = np.array([1, 1, 1])

print(a + b)
print(a - b)
print(a*b)
print(b/a)
print(a**2)

[2 3 4]
[0 1 2]
[1 2 3]
[1.         0.5        0.33333333]
[1 4 9]


In [25]:
# matrices (2d arrays)
A = np.array([[1, 0, 0],
              [0, 1, 0],
              [0, 0, 1]])
# equivalent
A = np.eye(3)
print(A)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [26]:
# since normal * is element-wise, matrix multiplication demands a separate function
c = np.dot(A, a)
print(c)

# OR if you are using Python 3.4+, as I recommend, there is a dedicated matrix multiplication operator...
c = A@a
print(c)

[1. 2. 3.]
[1. 2. 3.]


* You can even have more than 2 dimensions (tensors) if you want
* There are dozens of other things numpy implements, some of which we've already seen (`mean`, `amax`, `amin`, etc.)
* Don't reinvent the wheel; look at the documentation if you need something

# Generating an ER network
* This model ($G(n,p)$) is parameterized by number of nodes $n$ and connection probability $p$
* We will implement this as a function that takes $n$, $p$ as arguments

In [27]:
def erdos_renyi(n, p):
    # Create an empty graph
    G = nx.Graph()
    
    # add nodes (integers between zero and n-1)
    G.add_nodes_from(range(0, n))
    
    # for all possible pairs of nodes, add a link with probability p
    for node1 in range(0, n):
        for node2 in xrange(node1 + 1, n):
            if np.random.uniform() < p:
                G.add_edge(node1, node2)
    return G

In [28]:
# More pythonic way
import itertools as it

def erdos_renyi(n, p):
    G = nx.Graph()
    nodes = range(n)
    G.add_nodes_from(nodes)
    G.add_edges_from(edge for edge in it.combinations(nodes, 2) if np.random.uniform() < p)
    return G

### Let's test to see if it works

In [47]:
G = erdos_renyi(10**3, 1e-2)

print("G has {0} nodes.".format(len(G)))
print("G has {0} edges.".format(G.size()))
degrees = [G.degree(node) for node in G]
print("Avg. degree of G is {0}.".format(np.mean(degrees)))

G has 1000 nodes.
G has 4956 edges.
Avg. degree of G is 9.912.


Do these statistics make sense for the $n$ and $p$ we provided?

You can test how long it takes to run something using the IPython "magic" commands `%time` or `%timeit` (the former runs something once, the latter does it a bunch of times and takes the average)

In [49]:
print("Varying n:")
%time erdos_renyi(10**2, 1e-3)
%time erdos_renyi(10**3, 1e-3)
%time erdos_renyi(10**4, 1e-3)

print("Varying p:")
%time erdos_renyi(10**2, 1e-4)
%time erdos_renyi(10**2, 1e-3)
%time erdos_renyi(10**2, 1e-2)

Varying n:
CPU times: user 3.56 ms, sys: 1.11 ms, total: 4.67 ms
Wall time: 4.04 ms
CPU times: user 280 ms, sys: 1.75 ms, total: 282 ms
Wall time: 281 ms
CPU times: user 27.8 s, sys: 38.5 ms, total: 27.8 s
Wall time: 27.9 s
Varying p:
CPU times: user 3.2 ms, sys: 432 µs, total: 3.63 ms
Wall time: 3.42 ms
CPU times: user 3 ms, sys: 161 µs, total: 3.16 ms
Wall time: 3.11 ms
CPU times: user 3.12 ms, sys: 346 µs, total: 3.46 ms
Wall time: 3.33 ms


Does the scaling of these running times make sense given how we implemented the algorithm? What is the running time, in Big-$\mathcal{O}$ notation, of our ER implementation in terns of the network size $N$? In terms of $p$?

# Breadth first search

In [31]:
# a deque is like a list, but it supports O(1) time insertion/removal at either end
from collections import deque

def bfs(G, source):
    """ return a dictionary that maps node-->distance for all nodes reachable
        from the source node, in the unweighted undirected graph G """
    # set of nodes left to visit
    nodes = deque()
    nodes.append(source)
    
    # dictionary that gives True or False for each node
    visited = {node:False for node in G}
    visited[source] = True
    
    # Initial distances to source are: 0 for source itself, infinity otherwise
    dist = {node: np.inf for node in G}
    dist[source] = 0
    
    # while (container) is shorthand for "while this container is not empty"
    while nodes:
        # take the earliest-added element to the deque (why do we do this instead of popright?)
        node = nodes.popleft()
        
        # visit all neighbors unless they've been visited, record their distances
        for nbr in G.neighbors(node):
            if not visited[nbr]:
                dist[nbr] = dist[node] + 1
                visited[nbr] = True
                nodes.append(nbr)
    return dist

# Components
As explained in the slides, we can use BFS mutiple times to get all the components

In [32]:
def components(G):
    """ return a list of tuples, where each tuple is the nodes in in a component of G """
    components = []
    
    nodes_left = set(G.nodes())
    while nodes_left:
        src = nodes_left.pop()
        dist = bfs(G, src)
        component = tuple(node for node in dist.keys() if dist[node] < np.inf)
        components.append(component)
        nodes_left = nodes_left - set(component)
    return components

Let's test it on the 100 node network we generated above

In [50]:
# get list of all components
C = components(G)

# sort the components by size in descending order
C = sorted(C, key=lambda c: len(c), reverse=True)

# print the lengths of the components
component_sizes = [len(c) for c in C]
print(component_sizes)

[1000]


# Hands-on
* Let's use what we've learned today to reproduce some of the results on pg. 16 of Chapter 3: Random Networks
* I need 4 groups: Subcritical, Critical, Supercritical, and Connected
* ***You can use the networkx function `fast_gnp_random_graph(n, p)` to generate your networks***

## Subcritical, Critical, and Supercritical groups
You will consider networks of average degree $k = 0.5, k=1$, and $k=2$ respectively. Choose connection probabilities accordingly. Each group must do the following:
* Generate one random network each of sizes $N=10^2, 10^3, 10^4, 10^5$, and $10^6$ using your group's average degree
* For each network, use the code above to get the connected components. 
* Following the lesson on plotting distributions in the last lecture, modify the code from the last lecture to
plot the distribution of the sizes of the connected components in log-log scale. Plot this for all the networks in the same figure using different colors. 
* Calculate the size of the largest component for each of the 5 networks. Are they giant components? Write new code to plot the largest component size as a function of N in semi-log scale (hint: use `plt.semilogx`)
* Compare the above two results with your expectation from the book

## Connected group
You will consider networks of average degree $k = 20$; choose connection probabilities accordingly. You must do the following:
* Generate one random network each of sizes $N=10^2, 10^3, 10^4, 10^5,$ and $10^6$ with this average degree.
* Use `%timeit` on BFS for the network of size $10^6$ (you can pick an arbitrary source node. Why is that?) to get the number of seconds it takes on your laptop to get single-source shortest paths. Using this, give an estimate of how long it would take to calculate the diameter of this network, which would require calculating shortest path lengths for ALL possible pairs.
* As an alternative, figure out what the following code is doing to calculate the diameter approximately. Explain to the class.
* Use this code to plot the "pseudo-diameter" as a function of N in semi-log scale (hint: use `plt.semilogx`)

In [37]:
import random

def pseudo_diameter(G):
    # diameter = infinity if not connected
    if not nx.is_connected(G):
        return np.inf
    
    # pick a node from G randomly
    nodes = list(G.nodes())
    u = random.choice(nodes)
    diam = 0
    while True:
        # what do you think this does? You are allowed to read the NetworkX documentation
        d = nx.single_source_shortest_path_length(G, u)
        
        # get farthest node from u & the corresponding distance
        farthest_node, d_max = max(d.items(), key=lambda item: item[1])
        if d_max <= diam:
            return diam
        else:
            u = farthest_node
            diam = d_max

# Don't reinvent the wheel

Now that you understand how some basic graph analysis algorithms work, you should never use them again and instead use the following off-the-shelf commands which are better written and faster.

## Graph generation

* ***`erdos_renyi_graph(n, p)`***   
Generate a random graph. More or less the same as we implemented above.
* ***`fast_gnp_random_graph(n, p)`***   
Much faster for sparse graphs

## Paths and path length
All of the below work on both `Graph` and `DiGraph` objects

* ***`has_path(G, source, dest)`***   
test to see if there is a path (of any length) in G from `source` to `dest`   
* ***`shortest_path(G, source, dest)`*** and ***`shortest_path_length(G, source, dest)`***   
former returns path as a sequence of nodes, latter only returns the length   
* ***`all_shortest_paths(G, source, dest)`***   
same as above, but gives ALL shortest paths   
* ***`single_source_shortest_path(G, source)`*** and ***`single_source_shortest_path_length(G, source)`***   
return dictionary `d` where `d[node]` is respectively, the shortest path/path length from `source` to `node`   
* ***`all_pairs_shortest_path(G)`*** and ***`all_pairs_shortest_path_length(G)`***   
return dictionary `d` where `d[node1][node2]` is as above   
* ***`dijkstra_path(G, source, dest)`*** and ***`dijkstra_path_length(G, source, dest)`***   
As above, but for weighted `Graph`/`DiGraph` objects   
* ***`single_source_dijkstra_path(G, source)`*** and ***`single_source_dijkstra_path_length(G, source)`***   
As above, but for weighted `Graph`/`DiGraph` objects  

## Searching
All of the below work on both `Graph` and `DiGraph` objects

* ***`bfs_tree(G, source)`***   
Return a Di/Graph representing the tree spanned by a breadth-first search starting at `source`   
* ***`dfs_tree(G, source)`***  
Same using depth-first search (gives same result)   
* ***`all_shortest_paths(G, source, dest)`***   
same as above, but gives ALL shortest paths   

## Connectedness (Undirected)
The below work only on `Graph` objects

* ***`is_connected(G)`***   
`True` or `False` depending on whether `G` is connected or not      
* ***`connected_components(G)`***     
Return a list of lists, where each sub-list contains the nodes in one component   
* ***`number_connected_components(G)`***      
Returns only the length of the list above   
* ***`connected_component_sugraphs(G)`***      
Returns a list of new `Graph` objects each representing a component of `G`   
* ***`node_connected_component(G, node)`***      
Return a list of the nodes in the component of `G` containing `node`   

## Connectedness (Strong and weak)
The commands below work only on `DiGraph` objects

Note: the `is_weakly_` versions are equivalent to first converting the DiGraph to undirected using G.undirected(), and then running the undirected equivalents above.

* ***`is_strongly_connected(G)`***   
* ***`strongly_connected_components(G)`***     
* ***`number_strongly_connected_components(G)`***      
* ***`strongly_connected_component_sugraphs(G)`***     


* ***`is_weakly_connected(G)`***   
* ***`weakly_connected_components(G)`***     
* ***`number_weakly_connected_components(G)`***      
* ***`weakly_connected_component_sugraphs(G)`***  

All are analogous to the undirected case   
  

# Networks with scale-free degree distributions
* Let's use the configuration model to generate networks with (approximately) scale-free degree distributions.
* `configuration_model` in NetworkX produces a MultiGraph. Why? We will change that into a regular graph.

Remember, the configuration model takes as input a desired degree sequence. It then spits out a network having that degree sequence. So first, we need to be able to randomly generate degrees following a power-law distribution.

In [38]:
def powerlaw_degree_sequence(n, gamma, k_min):
    """ Implements the method for generating power-law distributed numbers
    from uniformly-distributed numbers described in Clauset et al., 2009,
    appendix D"""
    r = np.random.uniform(0, 1, size=n)
    deg = np.floor((k_min-0.5)*(1.0 - r)**(-1.0/(gamma-1)) + 0.5)
    deg = list(map(int, deg))
    return deg

In [39]:
def sf_net(n,  gamma, k_min):
    deg = powerlaw_degree_sequence(n, gamma, k_min)
    # sum of all degrees must be even. Why is that?
    if sum(deg) % 2 == 1:
        deg[0] += 1
    G = nx.configuration_model(deg)
    H = nx.Graph(G)
    H.remove_edges_from(H.selfloop_edges())
    return H

# Hands-on exercise
* Split into four groups, one each for $\gamma = 2.001, 2.5, 3$, and $3.5$
* All groups should use the above code to generate networks of sizes $10^2 \ldots 10^5$ as before with their chosen scaling exponent. Generate all networks with minimum degree cutoff $k_{min}$= 5

* First, measure the maximum degree of each network $k_{max}$, and then plot kmax in log-log scale as a function of $N$ 
* Next, I want you to plot the average shortest-path distance as a function of $N$ in semi-log scale. 
* Note that for larger networks it will be impossible to measure all pairs shortest paths. As an approximation, you should take a random *sample* of pairs of nodes (src, dest), measure the shortest path length between src and dest, and then take the average. Use 100 random node pairs per network.

* Hint 1: `[np.random.choice(G, size=2, replace=False) for _ in range(100)]` will give you a list of 100 random
node pairs from G
* Hint 2: You will need to run this within one component. Choose the largest. The following code will sort the components from largest to smallest

`components = sorted(components, key=len, reverse=True)`

You can then use the `subgraph` command on the first component (`components[0]`)
* Hint 3: Use `nx.shortest_path_length` and `nx.connected_components`. They are faster than what we've written.